### ISSUE TITLE
Is it possible to "De-compile" TorchCircuit?

### ISSUE THREAD

**Issue by Meaquadddd** [april-tools/cirkit#386]
Hi, thank you for creating and maintaining such an amazing library.

I came across several questions while using your library. It would be great if any help on them is provided.

I have a discriminative circuit $f$ and a generative circuit $p$ on the same scope $\mathrm{x}$, and they admit the same tree-shaped region graph. They are compiled and learnt with different loss functions, now I want to compute the expectation query
$\mathbb{E}_{\mathrm{x}\sim p(\mathrm{x})}\left[f(\mathrm{x})\right]$.

This is easy when they are both `SymbolicCircuit one can directly call the `multiply` query. However, after compilation, they became `TorchCircuit` and do not support the `multiply` query for `SymbolicCircuit`, is there any way to "decompile" this `TorchCircuit to a `SymbolicCircuit`? In my understanding, it should be able to do that by sharing the weight of the `layers` in `TorchCircuit` with `SymbolicCircuit`.

The second question shares the same setting as the first one. Now we assume that $\mathrm{x}$ is decomposed into two disjoint components $\mathbf{y}$ and $\mathrm{z}$, that is $\mathrm{x}=\mathbf{y}:\mathrm{z}$. I am interested in the following query
$\int_{\mathrm{z}}\left( \int_{\mathrm{y}} f(\mathrm{y};\mathrm{z})p(\mathrm{y}\left|\mathrm{z} \right.) \right) ^{2}p(\mathrm{z})$

Is it possible to induce a conditional circuit $p(\mathrm{y}\left|\mathrm{z} \right.)$ from the existing circuit $p(\mathrm{x})$?

Any help would be greatly appreciated !

**Comment by Meaquadddd** [comment]
The first question can be done with the following procedure 
```
from cirkit.pipeline import PipelineContext, compile

with PipelineContext.from_default_backend() as ctx:
    torch_circuit = compile(circuit)         

# your training process here
    
sc = ctx.get_symbolic_circuit(torch_circuit)
```

**Comment by loreloc** [comment]
Hi Xiliang,
Sorry for the late reply!

>In my understanding, it should be able to do that by sharing the weight of the layers in TorchCircuit with SymbolicCircuit.

Internally the compiler in the pipeline context keeps track of a map between symbolic parameters and compiled tensor parameters (in short, torch.Tensor objects). This is for all circuits compiled in that context. Note that in you code snippet ```sc``` and ```circuit``` are actually the same symbolic circuit instance, even after training. This is because the symbolic circuit does not hold parameters at all. As a consequence, you do not need to "get" the symbolic circuit after training. Instead, you could directly use the symbolic circuit ```circuit``` for all the symbolic operators, even after training. Upon compilation, the trained parameters will be referenced even if you operate on the symbolic circuit ```circuit``` you started with.

>Is it possible to induce a conditional circuit p ( y | z ) from the existing circuit p ( x ) ?

There is no symbolic conditioning of that kind at the moment, meaning that you cannot realize $$p(Y \mid Z)$$ as another circuit and operate on it to integrate the variables $$Z$$ later. So, I am afraid the closest thing you could do is use the ```evidence``` symbolic operator to set evidence for $$Z$$ and obtain a circuit $$q(Y) = p(Y, Z=z)$$. This means you might only be able to compute the inner integral $$\int f(Y,Z) p(Y\mid Z)^2 dY$$ given an assignment to $$Z$$.

Let me know if you have any other questions.
